src/librustc_trans/trans/closure.rs

```diff
@@ -150,8 +150,9 @@ pub fn get_or_create_closure_declaration<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
         return llfn;
     }
 
-    let path = ccx.tcx().def_path(closure_id);
-    let symbol = mangle_internal_name_by_path_and_seq(path, "closure");
+    let symbol = symbol_names::exported_name(ccx,
+                                             mono_id.def,
+                                             mono_id.params.as_slice());
```

Comment:
@eddyb 

> although I'm not sure whether my PR or this should come first 

It's a race! No, but seriously -- do you think they'll conflict much?

Also, I was just looking at callers of `exported_name`. The code in `get_item_val` seems a bit dubious. Not *wrong*, but dubious -- it calls `exported_name` with the `id` of (e.g.) an enum or function but a single type `T` where `T` is basically `lookup_item_type(id).ty`. This seems to ... mostly be used only for non-generic items, but in that case the type parameter should be altogether unnecessary I would think. Not sure if you refactored this at all? I remember some changes in that direction, so I suspect so. Seems like taking a `mono_id` would force the question and clean it up.